pkgs/build-support/mkshell/minimal.nix

```diff
@@ -0,0 +1,47 @@
+{ writeTextFile, writeScript, bash }:
+
+# A special kind of derivation that is only meant to be consumed by the
+# nix-shell. This differs from the traditional `mkShell` in that it is
+# ultra minimalist in dependencies.
+# It does not come with traditional stdenv (i.e. coreutils).
+# Its only dependency is essentially bash.
+{
+# a list of packages to add to the shell environment
+# we simplify here rather than `native` or `propagated` since
+# this is only ever used to be in a nix-shell; which is by default
+# the current system
+packages ? [ ], ... }@attrs:
+
+derivation ({
+  name = "minimal-nix-shell";
+
+  # What is the need for this reference to stdenv?
+  # Well Nix itself, checks to see if this attribute exists
+  # and sources the file.
+  # We will setup our nix-shell environment here and do things
+  # like clear the $PATH
+  # reference: https://github.com/NixOS/nix/blob/94ec9e47030c2a7280503d338f0dca7ad92811f5/src/nix-build/nix-build.cc#L494
+  "stdenv" = writeTextFile rec {
+    name = "setup";
+    executable = true;
+    destination = "/${name}";
+    text = ''
+      set -e
+      unset PATH
+
+      for a in $packages; do
+        export PATH=$a/bin:$PATH
+      done
```

Comment:
You'll also want to set `XDG_DATA_DIRS`, because that's what bash uses to find completion scripts on demand. Stdenv sets it generically since https://github.com/NixOS/nixpkgs/pull/103501, which inspired [devShell](https://github.com/numtide/devshell) to do the same.

Reply:
It comes at a very very little cost, but does enable shell completions.

If that's not the design choice you want to make, I wonder if it should even set PATH. Maybe someone only wants to set another environment variable, in which case PATH is redundant.

I guess it boils down to the question "when is it not a shell anymore?" which is obviously not clearly delineated, but needs an answer for this to be actually _minimal_.

Reply:
Also, in what regard is it minimal? I would define it as the combination of the number of store paths and the total byte size. The latter is generally so disproportionally determined by dependencies that the size of the shellHook (or similar) does not matter, as long as it doesn't increase store path count.
I do consider store path count separately because it introduces latency with nix-copy-closure and most binary caches.

Reply:
I would say minimal with regards to closure size.